docs/src/rust-client/oracle_tutorial.md

```diff
@@ -0,0 +1,360 @@
+
+```bash
+cargo new miden-defi-app
+cd miden-defi-app
+```
+
+Add the following dependencies to your `Cargo.toml` file:
+
+```toml
+miden-client = { version = "0.8.1", features = ["testing", "concurrent", "tonic", "sqlite"] }
+miden-lib = { version = "0.8", default-features = false }
+miden-objects = { version = "0.8", default-features = false }
+miden-crypto = { version = "0.14.0", features = ["executable"] }
+miden-assembly = "0.13.0"
+rand = { version = "0.9" }
+serde = { version = "1", features = ["derive"] }
+serde_json = { version = "1.0", features = ["raw_value"] }
+tokio = { version = "1.40", features = ["rt-multi-thread", "net", "macros"] }
+rand_chacha = "0.9.0"
+miden-client-tools = "0.1.0"
+```
+
+### Step 1: Set up your `src/main.rs` file
+
+Copy and paste the following code into your `src/main.rs` file:
+
+```rust
+use rand::RngCore;
+use std::{fs, path::Path};
+
+use miden_client::{
+    account::{AccountBuilder, AccountId, AccountStorageMode, AccountType, StorageSlot, component::AccountComponent},
+    rpc::{
+        domain::account::{AccountStorageRequirements, StorageMapKey},
+        Endpoint,
+    },
+    transaction::{
+        ForeignAccount, TransactionKernel, TransactionRequestBuilder, TransactionScript,
+    },
+    Client, ClientError, Felt, Word, ZERO,
+};
+
+use miden_client_tools::{
+    instantiate_client, create_library, delete_keystore_and_store,
+};
+
+/// Import the oracle + its publishers and return the ForeignAccount list
+pub async fn get_oracle_foreign_accounts(
+    client: &mut Client,
+    oracle_account_id: AccountId,
+    trading_pair: u64
+) -> Result<Vec<ForeignAccount>, ClientError> {
+    client.import_account_by_id(oracle_account_id).await?;
+
+    let oracle_record = client
+        .get_account(oracle_account_id)
+        .await
+        .expect("RPC failed")
+        .expect("oracle account not found");
+
+    let storage = oracle_record.account().storage();
+    let publisher_count = storage.get_item(1).unwrap()[0].as_int();
+
+    let publisher_ids: Vec<AccountId> = (1..publisher_count.saturating_sub(1))
+        .map(|i| {
+            let digest = storage.get_item(2 + i as u8).unwrap();
+            let words: Word = digest.into();
+            AccountId::new_unchecked([words[3], words[2]])
+        })
+        .collect();
+
+    let mut foreign_accounts = Vec::with_capacity(publisher_ids.len() + 1);
+
+    for pid in publisher_ids {
+        client.import_account_by_id(pid).await?;
+
+        foreign_accounts.push(ForeignAccount::public(
+            pid,
+            AccountStorageRequirements::new([(
+                1u8,
+                &[StorageMapKey::from([
+                    ZERO,
+                    ZERO,
+                    ZERO,
+                    Felt::new(trading_pair),
+                ])],
+            )]),
+        )?);
+    }
+
+    foreign_accounts.push(ForeignAccount::public(
```

Comment:
Why do you need to add the oracle id itself to the list of foreign accounts? 

The list contains all foreign accounts + the oracle Id, right? Do you call all those accounts foreign accounts?